lib/logproto/logproto-proxied-text-server.c

```diff
@@ -189,31 +189,100 @@ _log_proto_proxied_text_server_add_aux_data(LogProtoProxiedTextServer *self, Log
+                                     &(self->proxy_header_buff[self->proxy_header_buff_len]),
+                                     sizeof(gchar), NULL);
+      if(rc < 0)
+        {
+          if (errno == EAGAIN)
+            return LPS_AGAIN;
+          else
+            {
+              msg_error("I/O error occurred while reading proxy header", evt_tag_int(EVT_TAG_FD,
+                        self->super.super.super.transport->fd),
+                        evt_tag_error(EVT_TAG_OSERROR));
+              return LPS_ERROR;
+            }
+        }
+
+      /* permissive termination */
+      if(rc == 0)
+        return LPS_SUCCESS;
+
+      self->proxy_header_buff_len++;
+      self->proxy_header_buff[self->proxy_header_buff_len] = '\0';
+      if (self->proxy_header_buff[self->proxy_header_buff_len - 1] == '\n')
+        {
+          return LPS_SUCCESS;
+        }
+
+    }
+
+  msg_error("PROXY proto header with invalid header length",
+            evt_tag_int("max_parsable_length", PROXY_PROTO_HDR_MAX_LEN),
+            evt_tag_int("max_length_by_spec", PROXY_PROTO_HDR_MAX_LEN_RFC),
+            evt_tag_long("length", self->proxy_header_buff_len),
+            evt_tag_str("header", (const gchar *)self->proxy_header_buff));
+  return LPS_ERROR;
+}
+
+static void
+_log_proto_proxied_text_server_switch_to_tls(LogProtoProxiedTextServer *self)
+{
+  if (!multitransport_switch((MultiTransport *)self->super.super.super.transport, transport_factory_tls_id()))
+    {
+      msg_error("proxied-tls failed to switch to TLS");
+      return;
+    }
+  msg_debug("proxied-tls switch to TLS: OK");
+}
+
 static LogProtoStatus
 _log_proto_proxied_text_server_handshake(LogProtoServer *s)
 {
-  const guchar *msg;
-  gsize msg_len;
-  gboolean may_read = TRUE;
-  gboolean parsable;
-  LogProtoStatus status;
-
   LogProtoProxiedTextServer *self = (LogProtoProxiedTextServer *) s;
 
-  // Fetch a line from the transport layer
-  status = log_proto_buffered_server_fetch(&self->super.super.super, &msg, &msg_len, &may_read, NULL, NULL);
+  LogProtoStatus status = _fetch_into_proxy_buffer(self);
 
   self->handshake_done = (status == LPS_SUCCESS);
   if (status != LPS_SUCCESS)
     return status;
 
-  parsable = _log_proto_proxied_text_server_parse_header(self, msg, msg_len);
+  gboolean parsable = _log_proto_proxied_text_server_parse_header(self, self->proxy_header_buff,
+                      self->proxy_header_buff_len);
 
-  msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s", (gint) msg_len, msg));
+  msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s",
+                                                             (gint) self->proxy_header_buff_len, self->proxy_header_buff));
 
   if (parsable)
     {
       msg_info("PROXY protocol header parsed successfully");
+
+      if (self->has_to_switch_to_tls)
+        _log_proto_proxied_text_server_switch_to_tls(self);
```

Comment:
I think we should close the connection if we fail to switch to TLS, otherwise, we'd start processing encrypted payload without decrypting it. syslog-ng handles almost everything in no-parse mode, so it could be a real problem :)

For example:
```
if (self->has_to_switch_to_tls && !_log_proto_proxied_text_server_switch_to_tls(self))
  return LPS_ERROR;
```